.github/workflows/tests.yml

```diff
@@ -62,11 +74,75 @@ jobs:
         key: ${{ runner.os }}-cypress-${{ hashFiles('apps/site/*/package-lock.json') }}
       id: cypress-cache
     - name: <node> Install dependencies (if needed)
-      if: steps.node-cache.outputs.cache-hit != 'true'
+      if: ${{ steps.node-cache.outputs.cache-hit != 'true' }}
       run: |
         git config --global url."https://github.com/".insteadOf ssh://git@github.com/
         npm run install:ci
 
+  build_app:
+    name: Build & cache app
+    runs-on: ubuntu-latest
+    needs: setup
+    outputs:
+      build-key: ${{ steps.cache-keys.outputs.build }}
+      digest-key: ${{ steps.cache-keys.outputs.digest }}
+      assets-key: ${{ steps.cache-keys.outputs.assets }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set cache keys
+        id: cache-keys
+        run: |
+          echo "::set-output name=digest::${{ runner.os }}-digest-cache-${{ hashFiles('apps/site/priv/static/**') }}"
+          echo "::set-output name=build::${{ runner.os }}-build-cache-${{ hashFiles('**/lib/**.ex', 'apps/site/priv/static/**', '**/mix.lock') }}"
+          echo "::set-output name=assets::${{ runner.os }}-assets-cache-${{ hashFiles('apps/site/assets/**', 'apps/site/react_renderer/**.js') }}"
+      - uses: actions/cache@v2
+        with:
+          path: ~/.asdf
+          key: ${{ needs.setup.outputs.asdf-key }}
+      - uses: mbta/actions/reshim-asdf@v1
+      - uses: actions/cache@v2
+        with:
+          path: deps
+          key: ${{ needs.setup.outputs.mix-key }}
+      - uses: actions/cache@v2
+        with:
+          path: |
+            apps/site/assets/node_modules
+            apps/site/react_renderer/node_modules
+          key: ${{ needs.setup.outputs.nodejs-key }}
+      
+      - name: Cache front end assets
+        uses: actions/cache@v2
+        with:
+          path: |
+            apps/site/priv/static/js
+            apps/site/react_renderer/dist/app.js
+          key: ${{ steps.cache-keys.outputs.assets }}
+        id: assets-cache
+      - run: npm run webpack:build
+        if: ${{ steps.assets-cache.outputs.cache-hit != 'true' }}
+        working-directory: apps/site/assets
+      - run: npm run build
+        if: ${{ steps.assets-cache.outputs.cache-hit != 'true' }}
+        working-directory: apps/site/react_renderer
+
+      - name: Digest assets # always, as the cache key hash needs these files
+        run: mix phx.digest
```

Comment:
I'm not sure this works: the hash of the files is happening in an earlier step, but the relevant files aren't created until this step. Can you see what happens if you change one of the CSS files which gets compiled into `apps/site/priv/static`?

Reply:
Ahh good point - I hadn't thought about the timing, but that's very true that since I'm running `hashFiles` in `step.cache-keys`, the `digest-key` has no files to hash yet. 🤔 And that probably explains why I saw my digest cache failing my expectations on earlier runs. 

I had decided to _always_ run `mix phx.digest` to get around that, but that's a very blunt approach and I can probably adjust the timing/sequence of operations here to get what I need. Time to revise....

Reply:
Simplified this slightly -- I now cache the `apps/site/priv/static` directory and use the cache key I was using for the assets (involving `hashFiles('apps/site/assets/**', 'apps/site/react_renderer/**.js')`). I figure that whenever I need to run the webpack build, I'll need to subsequently re-run `mix phx.digest` as well.